### PR TITLE
Simplify MatchCollection's ICollection.CopyTo

### PR DESCRIPTION
**Pull request by justinvp**
- Uses [`List<T>`'s implementation of `ICollection.CopyTo`](https://github.com/Microsoft/referencesource/blob/9da503f9ef21e8d1f2905c78d4e3e5cbb3d6f85a/mscorlib/system/collections/generic/list.cs#L394-L407) which is behavior compatible with the previous implementation.
- Avoids an unnecessary array allocation by no longer using
`List<T>.ToArray()`.
- `SR.Arg_RankMultiDimNotSupported` and `SR.Arg_InvalidArrayType` are no
longer needed.
- Avoids accessing a property to trigger a side effect.

### CODE REVIEW

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexMatchCollection.cs [hunk 1]
```diff
@@ -134,22 +134,10 @@ bool ICollection.IsSynchronized
         /// </summary>
         void ICollection.CopyTo(Array array, int arrayIndex)
         {
-            if ((array != null) && (array.Rank != 1))
-            {
-                throw new ArgumentException(SR.Arg_RankMultiDimNotSupported);
-            }
+            if (!_done)
+                GetMatch(s_infinite);
```
Comment (sharwell) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexMatchCollection.cs]: I would prefer either of the following to this change:

1. Keep the reference to `Count` (with the associated comment)
2. Refactor this conditional call to `GetMatch` to its own method (e.g. `LazyInit` or something), and call the new method from both the `Count` and this method

**Comment by sharwell** [comment]
6492e17 LGTM, but I have reservations about 130acb9.

**Comment by justinvp** [comment]
Alright, I moved this to a separate method, called from `Count` and `CopyTo`.

**Comment by sharwell** [comment]
:+1:

**Pull request merged**